### PR TITLE
JS - Serialize/Deserialize game state

### PR DESCRIPTION
**Pull request by bitblip**
- `F5` to quick-save
- `F9` to quick-load

You can quick-load from the main menu

### CODE REVIEW

**Comment by bitblip** [comment]
I had to rework the map widget a little to be compatible with saving. The trick is to look to PlayerState for map state.

File: MonoZelda/UI/HUDMapWidget.cs [hunk 1]
```diff
@@ -54,6 +54,9 @@ public override void Load(ContentManager content)
 
     public override void Update()
     {
+        SetCompassMarkerVisible(PlayerState.HasCompass);
+        SetMapVisible(PlayerState.HasMap);
+
```
Comment (xsedim) [inline MonoZelda/UI/HUDMapWidget.cs:59]: These should ideally only be called when the value changes, not every single frame, but they're small methods so this is not much of a performance hit.

File: MonoZelda/Save/SaveState.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        public HashSet<(string, DoorDirection)> Keyring { get; set; }
+        public HashSet<Point> DiscoveredRooms = new();
+
+        public ProjectileType EquipedProjectile { get; set; }
+
+        // Dungeon state
+
+        public string RoomName { get; set; }
+        public Dictionary<string, DungeonRoom> Rooms { get; set; }
+    }
```
Comment (xsedim) [inline MonoZelda/Save/SaveState.cs:33]: I believe this framework will be good for maintainability make it easy to add more saveable data in the future (RPG elements for sprint 5 like player level and XP).

**Review by xsedim — CHANGES_REQUESTED** [review]
Overall great work in implementing this feature which required a good deal of refactoring and keeping good code quality in mind.

One issue I found: saving from the starting room does not correctly store Link's starting position. To replicate, start the game, quick save without leaving the starting room, and then quick load from any room. I believe it is because Link's starting position is stored based on the door he entered a room from, but he never enters a door before saving so his position is set to (0, 0) outside the room.

Otherwise, the following features are correctly saved and loaded from my testing:
- Starting room
- Items (rupees, keys, bombs)
- Opened doors (key doors, diamond doors, bombable walls)
- Killed enemies
- Collected items (map, compass, keys, potion, boomerang, bow, triforce)
- Player health
- Equipped projectile
- Player map marker

**Review by xsedim — APPROVED** [review]
Tested — looks good

**Pull request merged**